pkg/controller/admissionchecks/provisioning/controller.go

```diff
@@ -833,3 +830,88 @@ func limitObjectName(fullName string) string {
 	hashBytes := hex.EncodeToString(h.Sum(nil))
 	return fmt.Sprintf("%s-%s", fullName[:objNameMaxPrefixLength], hashBytes[:objNameHashLength])
 }
+
+type MergedPodSet struct {
+	Name             kueue.PodSetReference
+	PodSet           *kueue.PodSet
+	PodSetAssignment *kueue.PodSetAssignment
+	Count            int32
+}
+
+func mergePodSets(
+	wl *kueue.Workload,
+	prcSpec *kueue.ProvisioningRequestConfigSpec,
+) ([]MergedPodSet, error) {
+	expectedPodSets := requiredPodSets(wl.Spec.PodSets, prcSpec.ManagedResources)
+	psaMap := slices.ToRefMap(wl.Status.Admission.PodSetAssignments, func(p *kueue.PodSetAssignment) kueue.PodSetReference { return p.Name })
+	podSetMap := slices.ToRefMap(wl.Spec.PodSets, func(ps *kueue.PodSet) kueue.PodSetReference { return ps.Name })
+
+	mergePolicy := prcSpec.PodSetMergePolicy
+	mergedPodSets := []MergedPodSet{}
+	for _, psName := range expectedPodSets {
+		ps, psFound := podSetMap[psName]
+		psa, psaFound := psaMap[psName]
+		if !psFound || !psaFound {
+			return nil, errInconsistentPodSetAssignments
+		}
+
+		// Try to merge with an existing pod set
+		merged := false
+		if mergePolicy != nil {
+			for i, mps := range mergedPodSets {
+				if merged = canMergePodSets(mps.PodSet, ps, mergePolicy); merged {
+					mergedPodSets[i].Count += ptr.Deref(psa.Count, ps.Count)
+					break
+				}
+			}
+		}
+
+		// If not merged (or if mergePolicy is nil), add as a new pod set
```

Comment:
Drop this comment, it seems not matching the impl, and it is also self-explanatory